lib/ast-to-react.js

```diff
@@ -44,33 +49,21 @@
  *
  * To do: is `data-sourcepos` typeable?
  *
- * @callback CodeComponent
- * @param {JSX.IntrinsicElements['code'] & ReactMarkdownProps & {inline?: boolean}} props
- * @returns {ReactNode}
- *
- * @callback HeadingComponent
- * @param {JSX.IntrinsicElements['h1'] & ReactMarkdownProps & {level: number}} props
- * @returns {ReactNode}
- *
- * @callback LiComponent
- * @param {JSX.IntrinsicElements['li'] & ReactMarkdownProps & {checked: boolean|null, index: number, ordered: boolean}} props
- * @returns {ReactNode}
- *
- * @callback OrderedListComponent
- * @param {JSX.IntrinsicElements['ol'] & ReactMarkdownProps & {depth: number, ordered: true}} props
- * @returns {ReactNode}
- *
- * @callback TableCellComponent
- * @param {JSX.IntrinsicElements['table'] & ReactMarkdownProps & {style?: Object.<string, unknown>, isHeader: boolean}} props
- * @returns {ReactNode}
- *
- * @callback TableRowComponent
- * @param {JSX.IntrinsicElements['tr'] & ReactMarkdownProps & {isHeader: boolean}} props
- * @returns {ReactNode}
+ * @typedef {JSX.IntrinsicElements['code'] & ReactMarkdownProps & {inline?: boolean}} CodeProps
+ * @typedef {JSX.IntrinsicElements['h1'] & ReactMarkdownProps & {level: number}} HeadingProps
+ * @typedef {JSX.IntrinsicElements['li'] & ReactMarkdownProps & {checked: boolean|null, index: number, ordered: boolean}} LiProps
+ * @typedef {JSX.IntrinsicElements['ol'] & ReactMarkdownProps & {depth: number, ordered: true}} OrderedListProps
+ * @typedef {JSX.IntrinsicElements['table'] & ReactMarkdownProps & {style?: Object.<string, unknown>, isHeader: boolean}} TableCellProps
+ * @typedef {JSX.IntrinsicElements['tr'] & ReactMarkdownProps & {isHeader: boolean}} TableRowProps
+ * @typedef {JSX.IntrinsicElements['ul'] & ReactMarkdownProps & {depth: number, ordered: false}} UnorderedListProps
```

Comment:
I'm curious on the thinking of creating dedicated prop types vs inlining this in the `ComponentType<>`.
Are these props something you do/plan to import directly?

Reply:
Yes, they would be very helpful for me to import directly. In my example:
```
export default function Markdown(props: ReactMarkdownOptions) {
  return (
    <ReactMarkdown
      components={{
        li: withStyles(styles)(
          ({
            classes,
            color,
            node,
            ref,
            ...props
          }: JSX.IntrinsicElements['li'] &
            ReactMarkdownProps & {
              checked: boolean | null;
              index: number;
              ordered: boolean;
            } & WithStyles<typeof styles>) => (
            <li className={classes.listItem}>
              <Typography paragraph component="span" {...props} />
            </li>
          ),
        ),
      }}
      {...props}
    />
  );
}
```

becomes:

```
export default function Markdown(props: ReactMarkdownOptions) {
  return (
    <ReactMarkdown
      components={{
        li: withStyles(styles)(
          ({
            classes,
            color,
            node,
            ref,
            ...props
          }: LiProps & WithStyles<typeof styles>) => (
            <li className={classes.listItem}>
              <Typography paragraph component="span" {...props} />
            </li>
          ),
        ),
      }}
      {...props}
    />
  );
}
```

Reply:
Does `LiProps & WithStyles<typeof styles>` need to be explicitly added?
Shouldn't this be inferred by TSC without an explicit annotation? :thinking:

Reply:
Unfortunately I think it does need to be explicit. All TypeScript sees is that the function is getting passed into `withStyles`. I don't think it's smart enough to then try to figure out how to conform to the type for the `components` prop.

Reply:
That is unfortunate.
Oh well, thanks for the additional context!